### PR TITLE
Edge-triggered epoll events

### PR DESCRIPTION
**Pull request by ppopth**
EDITED: I have already implemented this feature for tcp sockets, udp sockets, unix domain sockets, pipes, eventfds.

### CODE REVIEW

File: src/main/host/descriptor/eventfd.rs
Comment (stevenengler) [inline src/main/host/descriptor/eventfd.rs:216]: This seems like it could lead to accidental bugs, so I think it would be nicer if this method took a bool flag, and only flips the `INPUT_BUFFER_PARITY` bit if the flag is true. This would force the caller to consider this behaviour, even if we currently always pass `true`. (And as mentioned in a different comment, the `INPUT_BUFFER_PARITY` doesn't really represent a state anyways, so I think handling it separately is okay.)

**Comment by stevenengler** [comment]
> The reason why you didn't want to integrate this as you mentioned in #2673 is that Shadow already "does the right thing", but, without this feature, I have to fork the upstream repo and convince the libp2p folks to use the forked repo instead.

Just to clarify, we would like to better follow what Linux does, but we just didn't have time to work on it (look in-depth into the Linux behaviour, consider the best way to design it, any performance implications, etc). If Shadow can better model the Linux behaviour, have a bunch of tests for it, and not significantly slow down the simulator, then I think it would be great to have this behaviour. IIRC the old PR didn't pass all of the tests, and Shadow's old C epoll code was a bit difficult to follow, so we also were more worried about introducing bugs. Now with the epoll code in rust, we can hopefully be more confident in any changes.

I'll leave a few comments that I have, but I'm not as knowledgeable about epoll so I'll let other people weigh in on this PR.

> I implemented it only for eventfd and legacy tcp

I think before merging we would want this to be consistent across Shadow for all file types, unless Linux does differently. And this is where tests would be useful, to make sure all of the file types behave this way. IIRC pipes also have this edge-triggered behaviour on Linux.

File: src/main/host/descriptor/socket.c
Comment (stevenengler) [inline src/main/host/descriptor/socket.c:412]: Is this true? For example if you read from a socket, will epoll always trigger an edge-triggered "readable" event? This is something that there should be a test for.

**Comment by stevenengler** [comment]
I'd expect that this causes more state updates, resulting in the event source running more callbacks, but the current version of this PR [doesn't seem](https://github.com/shadow/benchmark-results/tree/master/tor/2023-12-07-T22-11-41) to effect tor simulation runtime performance.

run_time

File: src/main/host/descriptor/mod.rs
Comment (ppopth) [reply src/main/host/descriptor/mod.rs:120]: Yeah, I understand that this is quite hacky. Since I think that this feature probably requires big architectural change and we don't know yet what it should be, I implemented the easiest and shortest possible approach.

File: src/main/host/descriptor/epoll/entry.rs
Comment (ppopth) [reply src/main/host/descriptor/epoll/entry.rs:81]: Ahh, not sure. I will dig it deeper when we decide to integrate this feature.

**Comment by ppopth** [comment]
> I think before merging we would want this to be consistent across Shadow for all file types, unless Linux does differently. And this is where tests would be useful, to make sure all of the file types behave this way. IIRC pipes also have this edge-triggered behaviour on Linux.

Totally agree, however, since this patch is quite hacky and we don't know yet how we should implement it. I will implement the rest when we know it, otherwise the effort will be wasted.

File: src/main/host/descriptor/eventfd.rs
Comment (ppopth) [reply src/main/host/descriptor/eventfd.rs:216]: Sounds like a good idea.

File: src/main/host/descriptor/socket.c
Comment (ppopth) [reply src/main/host/descriptor/socket.c:412]: Good point. Not sure either. I will check that again.

Comment (ppopth) [reply src/main/host/descriptor/socket.c:412]: No, it won't. `INPUT_BUFFER_PARITY` is used only to clear the `collected` field of the epoll entry. See https://github.com/shadow/shadow/blob/fb1072c/src/main/host/descriptor/epoll/entry.rs#L77-L81

So the event will be triggered only if the file is readable.

Apologize that I don't have the test for this yet.

**Comment by ppopth** [comment]
@robgjansen @stevenengler Are you okay with the `FileState::INPUT_BUFFER_PARITY`? I feel that it's probably not the most proper way to do, but I guess the most proper way may require a big change.

If you are okay to merge it with that, I will do all the necessary work on that to get it merged: fuzz tests, other tests, and edge-triggered event for all the file types.

My current feeling is that, if we don't want `FileState::INPUT_BUFFER_PARITY` and the proper way is too big, all the effort will be wasted.

Honestly, I'm not sure if the proper way will be really too big, I just feel that it will be. Let me think a moment to find a better way with small change.

**Comment by stevenengler** [comment]
> @robgjansen @stevenengler Are you okay with the FileState::INPUT_BUFFER_PARITY? I feel that it's probably not the most proper way to do, but I guess the most proper way may require a big change.

I think my opinion is that it's okay. In that case the flag should have a comment explaining why it's done this way, and that it's an exception. Alternatively, what do you think of the design in #3250? It adds a new argument for listeners so that files can emit signals and listeners can opt to receive them. It might be overkill for this, but we might want this ability to emit other types of signals in the future as well.

**Comment by robgjansen** [comment]
I would prefer proper signal listener support from #3250 over using `FileState::INPUT_BUFFER_PARITY` as an exceptional case. If #3250 is merged, then this PR becomes much easier to implement using the new signaling features from #3250.

**Comment by stevenengler** [comment]
We've now merged #3250, so it should be ready to use in this PR.

So what we're looking for before we can merge this PR is:

- use the new signal mechanism instead of the `INPUT_BUFFER_PARITY` state
- use the new edge-triggered epoll behaviour in all file types (udp, pipe, etc) and not just eventfd/tcp, as long as linux does as well (the tests should show which files do or don't)
- does this edge-triggered behaviour occur only on the input buffer, or also the output buffer?
- for the input buffer, does the edge-triggered behaviour occur when either reading or writing from/to the buffer? (Does reading from the buffer also trigger epoll?)
- new syscall tests in rust that test edge-triggered epoll under these various conditions for different file types to make sure that shadow behaves the same as linux

Thanks for working on this!

#### Tests

We have some existing C epoll tests in `src/test/epoll/test_epoll.c` and Rust epoll tests in `src/test/epoll/test_epoll.rs`. The new tests should be written in rust, but it's up to you whether they should go in `test_epoll.rs` or a new file. If the tests are large, it might make sense to have these edge-triggered tests in a new file.

You can test the epoll syscalls using the libc library (ex: [`libc::epoll_ctl`](https://docs.rs/libc/latest/libc/fn.epoll_ctl.html)). In `test_epoll.rs` we used the nix library to test the syscalls, but I think that was a mistake and we should have used the libc or linux-api libraries instead. The tests are also stuck on an old version of nix see (#3141) which uses an old version of their epoll API.

For us as maintainers, we don't really know exactly how linux behaves with this edge-triggered-but-not-really-edge-triggered epoll, so the the more tests you add the easier it will be to convince us that the changes accurately follow linux and that it's good to merge.

**Comment by ppopth** [comment]
> * Does reading from the buffer also trigger epoll?

From the test I implemented in https://github.com/shadow/shadow/issues/3274, the answer is no in Linux. However, it's surprising that the answer is yes in the current main branch of Shadow.

Since I don't want to pay the effort fixing this and I don't know any application depending on that behaviour, I will not add such the test in this PR.

**Comment by stevenengler** [comment]
> Since I don't want to pay the effort fixing this and I don't know any application depending on that behaviour, I will not add such the test in this PR.

Thanks for making that issue and writing a test case! If you want to, you can add the test to Shadow but in the `ShadowTest::new("write-then-read", ...)` use `set![TestEnvironment::Libc]` (and leave a comment referring to #3274). This indicates that the test is supposed to pass in Shadow but doesn't, and you don't need to fix it since it will be skipped when running in Shadow.

**Comment by ppopth** [comment]
@stevenengler What is the state of `SharedBuf`? Is it something you still want to use in the long term?

In order to implement edge-triggered epoll for pipe and unix sockets, it seems like I have to implement `BufferSignals` (just like `FileSignals`) in addition to `BufferState` for `SharedBuf`.

If you are planning to deprecate `SharedBuf`, it's probably not worth the effort. If no, I will do the `BufferSignals`.

**Comment by ppopth** [comment]
> does this edge-triggered behaviour occur only on the input buffer, or also the output buffer?

I found strange behaviour on the output buffer. This edge-triggered thing occurs on the output buffer of udp sockets, but not tcp sockets or unix datagram sockets. I think it will take too much effort doing research on this, so I will not add the tests for those cases.

**Comment by stevenengler** [comment]
> What is the state of `SharedBuf`? Is it something you still want to use in the long term?

I think we still want it in the long term. We need some kind of shared buffer between two files (or sometimes more if there are multiple readers or writers).

> In order to implement edge-triggered epoll for pipe and unix sockets, it seems like I have to implement `BufferSignals` (just like `FileSignals`) in addition to `BufferState` for `SharedBuf`.
>
> If you are planning to deprecate `SharedBuf`, it's probably not worth the effort. If no, I will do the `BufferSignals`.

I forgot about the `SharedBuf` when adding the signals. I made #3276 to add the boilerplate code needed to support signals in `SharedBuf`. I don't want to conflict with the code you're working on, so if you've already started adding signals to `SharedBuf` let me know and we can use your code instead. If not, hopefully #3276 will make this PR easier.

> I found strange behaviour on the output buffer. This edge-triggered thing occurs on the output buffer of udp sockets, but not tcp sockets or unix datagram sockets. I think it will take too much effort doing research on this, so I will not add the tests for those cases.

Okay thanks for checking. I think it's okay to leave the output buffer out of the scope of this PR then. But please open an issue so we can remember that this edge-triggered behaviour doesn't follow linux for all output buffers and that it needs more investigation.

**Comment by ppopth** [comment]
@stevenengler What is your preferred way to do `write` to a raw file descriptor in the tests?
1. Use `std::fs::File::from_raw_fd` just like in https://stackoverflow.com/questions/54858018/how-do-i-write-to-a-specific-raw-file-descriptor-from-rust
2. or, use `libc::write` just like https://github.com/shadow/shadow/blob/348bc4/src/test/pipe/test_pipe.rs#L185-L198

**Comment by stevenengler** [comment]
> @stevenengler What is your preferred way to do `write` to a raw file descriptor in the tests?
> 
> 1. Use `std::fs::File::from_raw_fd` just like in https://stackoverflow.com/questions/54858018/how-do-i-write-to-a-specific-raw-file-descriptor-from-rust
> 
> 2. or, use `libc::write` just like https://github.com/shadow/shadow/blob/348bc4/src/test/pipe/test_pipe.rs#L185-L198
> 
> 
> 
> In test_epoll.rs, it uses `nix::unistd::write`.

The preferred is option (2).

In the tests, if it's a syscall that you're specifically testing, then use plain `libc`. But you can use either `libc` or `nix` for other syscalls that you're not testing. I'd avoid using `std` functions in tests since it's difficult to know exactly what syscalls they're making.

So in `test_epoll.rs` since you're testing epoll related syscalls, use `libc` for the epoll syscalls (`libc::epoll_wait`), and you can use `nix` or `libc` for other syscalls like `read` or `write` (`libc::write` or `nix::unistd::write`).

**Comment by ppopth** [comment]
@stevenengler it's ready for your eyes

File: src/main/host/descriptor/epoll/entry.rs
Comment (stevenengler) [inline src/main/host/descriptor/epoll/entry.rs:1]: Can you add an integration test to `test_epoll_edge.rs` that uses `EPOLLONESHOT`? I think something like the following might be good:

1. `EPOLL_CTL_ADD` the descriptor using `EPOLLIN | EPOLLET | EPOLLONESHOT`
2. in a thread, do `epoll_wait()`
3. push data to the socket so that it becomes readable (and the `epoll_wait()` should return)
4. in the thread, do another `epoll_wait()`
5. push more data to the socket (the `epoll_wait()` should *not* return because epoll was not rearmed)
6. rearm the event using `EPOLL_CTL_MOD`
7. push more data to the socket (and the `epoll_wait()` should return)

**Review by stevenengler — COMMENTED** [review]
It's looking good! I still have a few more files to look through (unix.rs, the rust tcp files, and only quickly looked through test_epoll_edge.rs), but thought I'd leave the comments I have so far. They're all minor comments, and two additional tests I think would be good to have.

File: src/test/epoll/test_epoll_edge.rs
Comment (stevenengler) [inline src/test/epoll/test_epoll_edge.rs:346]: This should work now because of the fix in #3277, so can you try using `all_envs.clone()`?

File: src/test/epoll/test_epoll_edge.rs
Comment (stevenengler) [inline src/test/epoll/test_epoll_edge.rs:107]: Can you rename this to `test_write_then_partial_read`, and then remove the test with the same name from `test_epoll.rs`?

**Review by stevenengler — CHANGES_REQUESTED** [review]
Looks good! All of the requested changes should be small, and then there are two additional tests I think would be good to have. After this I expect we can merge it.

> - use the new signal mechanism instead of the INPUT_BUFFER_PARITY state

Done.

> - use the new edge-triggered epoll behaviour in all file types (udp, pipe, etc) and not just eventfd/tcp, as long as linux does as well (the tests should show which files do or don't)

Done.

> - does this edge-triggered behaviour occur only on the input buffer, or also the output buffer?

@ppopth Once this is merged, can you make an issue for supporting the UDP output buffer behaviour you observed?

> - for the input buffer, does the edge-triggered behaviour occur when either reading or writing from/to the buffer? (Does reading from the buffer also trigger epoll?)

Done. There was a related bug #3274 where a partial read would trigger edge-triggered epoll in shadow but not Linux. This was fixed in #3277. This current PR contains a test that checks this for all socket and pipes.

> - new syscall tests in rust that test edge-triggered epoll under these various conditions for different file types to make sure that shadow behaves the same as linux

Done. The new tests check tcp, udp, unix sockets (all three types), pipes (stream and packet modes), and eventfds.

**Comment by stevenengler** [comment]
> #### Conflicting files
> src/main/host/descriptor/socket.c

In #3277 I moved the `legacyfile_adjustStatus((LegacyFile*)socket, STATUS_FILE_READABLE, TRUE, 0)` from socket.c to tcp.c, so you'll need to move the signal code there as well.

File: src/main/host/descriptor/socket/inet/udp.rs
Comment (ppopth) [reply src/main/host/descriptor/socket/inet/udp.rs:167]: Never mind. I think I just forgot to remove the comment.

File: src/main/host/descriptor/mod.rs
Comment (ppopth) [reply src/main/host/descriptor/mod.rs:150]: Yeah, that makes sense.

**Comment by ppopth** [comment]
> > #### Conflicting files
> > src/main/host/descriptor/socket.c
> 
> In #3277 I moved the `legacyfile_adjustStatus((LegacyFile*)socket, STATUS_FILE_READABLE, TRUE, 0)` from socket.c to tcp.c, so you'll need to move the signal code there as well to fix the conflict.

I resolved the conflict by merging instead of rebasing so that I would not lose the history. I will rebase it later.

File: src/main/host/descriptor/eventfd.rs
Comment (ppopth) [reply src/main/host/descriptor/eventfd.rs]: Done in 30555a269c3b9bf16efdf6346900870511fb2da6

File: src/main/host/descriptor/socket/unix.rs
Comment (ppopth) [reply src/main/host/descriptor/socket/unix.rs:1722]: I think it's not the case for `ConnLessInitial`

**Comment by ppopth** [comment]
@stevenengler it's ready for your eyes

**Comment by stevenengler** [comment]
> @stevenengler it's ready for your eyes

Looks good! Just do a rebase, then I'll approve it, run the benchmark, and merge it.

**Comment by ppopth** [comment]
@stevenengler it's all done. Can you please add a milestone to this PR?

**Review by stevenengler — APPROVED** [review]

**Comment by stevenengler** [comment]
[No change](https://github.com/shadow/benchmark-results/blob/master/tor/2024-01-24-T23-48-59/README.md) to the tor benchmark. I don't think tor uses edge-triggered epoll, so this should be expected.

run_time
transfer_time_5242880 exit

**Comment by stevenengler** [comment]
Thanks @ppopth!

**Comment by ppopth** [comment]
> @ppopth Once this is merged, can you make an issue for supporting the UDP output buffer behaviour you observed?

The issue is created at https://github.com/shadow/shadow/issues/3295

**Pull request merged**
